iamclient/iamclient.go

```diff
@@ -474,102 +512,128 @@ func (client *Client) ResumeNode(nodeID string) error {
+	client.identService = pb.NewIAMPublicIdentityServiceClient(client.publicConnection)
+	client.publicNodesService = pb.NewIAMPublicNodesServiceClient(client.publicConnection)
 
-		secureOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+	client.protectedConnection, err = grpchelpers.CreateProtectedConnection(client.config.CertStorage,
+		client.config.IAMProtectedServerURL, iamRequestTimeout, client.cryptocontext, client, client.insecure)
+	if err != nil {
+		return aoserrors.Wrap(err)
 	}
 
-	if connection, err = grpc.DialContext(ctx, serverURL, secureOpt, grpc.WithBlock()); err != nil {
-		return nil, aoserrors.Wrap(err)
+	client.certificateService = pb.NewIAMCertificateServiceClient(client.protectedConnection)
+	client.provisioningService = pb.NewIAMProvisioningServiceClient(client.protectedConnection)
+	client.nodesService = pb.NewIAMNodesServiceClient(client.protectedConnection)
+
+	// Restore GRPC NodeInfo subscription
+	if err = client.subscribeNodeInfoChange(); err != nil {
+		return aoserrors.Wrap(err)
 	}
 
-	return connection, nil
+	return nil
 }
 
-func (client *Client) getNodeInfo() (nodeID, nodeType string, err error) {
-	ctx, cancel := context.WithTimeout(context.Background(), iamRequestTimeout)
-	defer cancel()
+func (client *Client) closeGRPCConnection() {
+	log.Debug("Closing IAM connection...")
 
-	response, err := client.publicService.GetNodeInfo(ctx, &empty.Empty{})
-	if err != nil {
-		return "", "", aoserrors.Wrap(err)
+	if client.publicConnection != nil {
+		client.publicConnection.Stop()
 	}
 
-	log.WithFields(log.Fields{
-		"nodeID":   response.GetNodeId(),
-		"nodeType": response.GetNodeType(),
-	}).Debug("Get node Info")
+	if client.protectedConnection != nil {
+		client.protectedConnection.Close()
+	}
 
-	return response.GetNodeId(), response.GetNodeType(), nil
+	client.WaitIAMPublicServiceClient()
+	client.nodeInfoSubs.stopWG.Wait()
 }
 
-func (client *Client) createProtectedConnection(
-	config *config.Config, cryptocontext *cryptutils.CryptoContext, insecureConn bool) (
-	connection *grpc.ClientConn, err error,
-) {
-	var secureOpt grpc.DialOption
+func (client *Client) processEvents() {
+	for {
+		select {
+		case <-client.closeChannel:
+			return
 
-	if insecureConn {
-		secureOpt = grpc.WithTransportCredentials(insecure.NewCredentials())
-	} else {
-		certURL, keyURL, err := client.GetCertificate(config.CertStorage, nil, "")
-		if err != nil {
-			return nil, aoserrors.Wrap(err)
-		}
+		case <-client.tlsCertChan:
+			client.Lock()
+			client.reconnect()
+			client.Unlock()
 
-		tlsConfig, err := cryptocontext.GetClientMutualTLSConfig(certURL, keyURL)
-		if err != nil {
-			return nil, aoserrors.Wrap(err)
+		case <-client.reconnectChannel:
+			client.Lock()
+			client.reconnect()
+			client.Unlock()
 		}
+	}
+}
 
-		secureOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+func (client *Client) reconnect() {
+	if client.disableReconnect.CompareAndSwap(false, true) {
```

Comment:
It is unclear why disableReconnect is required. Need to discuss.

Reply:
disable reconnection attempts if any in progress